### PR TITLE
Make PUT on /data/fs idempotent.

### LINKED ISSUE #727: Attempting to PUT to an existing file throws an error
At least I think that's what the problem is:

```
{  
   "errors":[  
      {  
         "detail":"com.mongodb.MongoCommandException: Command failed with error -1: 'target namespace exists' on server localhost:27001. The full response is { \"ok\" : 0.0, \"errmsg\" : \"target namespace exists\" }"
      }
   ]
}
```

This occurs when attempting to put the `index` file for a notebook after the initial notebook has been created.

### PR DESCRIPTION
**Pull request by sellout**
Fixes #727.

### CODE REVIEW

**Comment by jdegoes** [comment]
Looks good, can you add an integration test so this issue doesn't pop up again?

File: web/src/main/scala/slamdata/engine/api/fs.scala [hunk 1]
```diff
@@ -356,7 +356,7 @@ class FileSystemApi(fs: FSTable[Backend]) {
 
     case req @ PUT -> AsPath(path) => for {
       body <- EntityDecoder.decodeString(req)
-      resp <- upload(body, path, (ds, p, json) => ds.save(p, json).attemptRun.leftMap(_ :: Nil))
+      resp <- upload(body, path, (ds, p, json) => ds.delete(p).flatMap(κ(ds.save(p, json))).attemptRun.leftMap(_ :: Nil))
```
Comment (mossprescott) [inline web/src/main/scala/slamdata/engine/api/fs.scala]: If `save` fails (like, say, if the input doesn't parse), the file is still deleted, right?

I think the semantics should be:
1. The file is is replaced with the new contents, or
2. The file is unchanged, and an error response is returned.

I don't think we can replace a collection atomically in MongoDB, so write-delete-rename is probably the best we can do. Not sure if you can do that with the current operations provided by `Filesystem`.

**Comment by sellout** [comment]
Yeah … I’m just not sure where/how to test this. Our integration tests don’t use the Web server, and our web tests use a `Stub` backend that doesn’t have this issue in the first place.

@mossprescott Any idea if there’s already somewhere to put tests for this, or should I build something new?

File: web/src/main/scala/slamdata/engine/api/fs.scala (hunk 1 above)
Comment (jdegoes) [inline web/src/main/scala/slamdata/engine/api/fs.scala]: > If save fails (like, say, if the input doesn't parse), the file is still deleted, right?

Good catch. We need to make sure if the save fails, then the original file is still there, which can be approximated with write-delete-rename.

**Comment by mossprescott** [comment]
This is why we need a free algebra for backend ops. The services would produce a description of the operations to be performed, and then we'd run an interpreter to produce the `Task[Response]`. That would make it easy to test all the "mutating" calls, which are basically not tested now.

If you need to add a method to `Filesystem` to support write-delete-rename, you could test that.

**Comment by sellout** [comment]
Uh, sorry … somehow I never saw the line comments … thought your message was just in response to the testing bit. But yeah, makes sense. I’ll get rid of the default and implement for Mongo properly.

**Comment by sellout** [comment]
Just noticed that the docs on `fs.save` claim that it replaces existing contents. Should I make that true, or should I change those docs?

**Comment by jdegoes** [comment]
Just for sake of completeness, we probably do want a "atomically create this file with this contents if it doesn't exist", as well as a "atomically create or replace the file with these contents whether or not they exist", and possibly a "atomically replace the file with these contents only if it exists", which would sort of be: `create`, `put`, and `replace`. Not sure how they map to existing FS operations, perhaps @mossprescott can comment more.

**Comment by jdegoes** [comment]
@sellout  BTW you should probably update the README docs while you're in there.

**Comment by sellout** [comment]
@jdegoes The README seems to describe the proper semantics, we just weren’t following them.

**Pull request merged**